corehq/reports.py

```diff
@@ -193,41 +196,78 @@ def _safely_get_report_configs(project_name):
     return configs
 
 
+def _make_report_class(config, show_in_dropdown=False, show_in_nav=False):
+    from corehq.apps.reports.generic import GenericReportView
+
+    # this is really annoying.
+    # the report metadata should really be pulled outside of the report classes
+    @classmethod
+    def get_url(cls, domain, **kwargs):
+        from corehq.apps.userreports.models import CUSTOM_REPORT_PREFIX
+        slug = (
+            ConfigurableReport.slug
+            if not config._id.startswith(CUSTOM_REPORT_PREFIX)
+            else CustomConfigurableReportDispatcher.slug
+        )
+        return reverse(slug, args=[domain, config._id])
+
+    @classmethod
+    def show_in_navigation(cls, domain=None, project=None, user=None):
+        if show_in_nav:
+            return config.visible or (
+            user and toggles.USER_CONFIGURABLE_REPORTS.enabled(user.username))
+        return False
+
+    @classmethod
+    def display_in_dropdown(cls, domain=None, project=None, user=None):
+        if show_in_dropdown:
+            return config.visible or (
+            user and toggles.USER_CONFIGURABLE_REPORTS.enabled(user.username))
+        return False
+
+    return type('DynamicReport{}'.format(config._id), (GenericReportView,), {
+        'name': config.title,
+        'description': config.description or None,
+        'get_url': get_url,
+        'show_in_navigation': show_in_navigation,
+        'display_in_dropdown': display_in_dropdown,
```

Comment:
Considering these are the same except for one variable, you could probably define the function just once, and use `partial`, something like so:
```python
from functools import partial

def mk_foobar_class(foo, bar):

    @classmethod
    def display_in(cls, domain=None, project=None, user=None, where=None):
        return (
            config.visible or (
                user and 
                toggles.USER_CONFIGURABLE_REPORTS.enabled(user.username)
            )
        ) if where else False

    return type('FooBar', (object,), {
        'show_in_foo': partial(display_in, where=foo),
        'display_in_bar': partial(display_in, where=bar),
    })
```